### PR TITLE
 FEATURE: FET-157 Upgrade to TypeScript 3

### PR DESCRIPTION
**Pull request by ivan-nejezchleb**
[Check PR owner responsibilities](https://confluence.intgdc.com/display/Development/Code-reviews#Code-reviews-Ownerresponsibilities)

Supported PR commands:

Command | Description
--- | ---
`extended test - examples` | Live examples tests
`extended test - storybook` | Storybook screenshot tests

See more [options](https://confluence.intgdc.com/display/kbhr/How+to+work+with+git+and+Github#HowtoworkwithgitandGithub-Extendedchecks).

# PR Checklist

- [x] Verify code changes ([Checklist](https://confluence.intgdc.com/display/Development/Code-reviews+checklist), [Best practices](https://confluence.intgdc.com/display/Development/Code-reviews+best+practices))
- [x] [Verify pull-request formalities](https://confluence.intgdc.com/display/Development/Code-reviews)
- [x] Change was tested by using [gdc-dev-release](https://confluence.intgdc.com/display/~tomas.vojtasek/Private+NPM) in [gdc-analytical-designer](https://github.com/gooddata/gdc-analytical-designer) and [gdc-dashboards](https://github.com/gooddata/gdc-dashboards) (if applicable)
- [x] Successful `extended test - examples`
- [x] Successful `extended test - storybook`
- [x] Checked yarn.lock consistency (no dep. duplicities especially Goodstrap)

# Related PRs

- gdc-app-components: https://github.com/gooddata/gdc-app-components/pull/544
- gdc-analytical-designer: https://github.com/gooddata/gdc-analytical-designer/pull/2189
- gdc-dashboards: https://github.com/gooddata/gdc-dashboards/pull/1884

# Related Jira tasks

### CODE REVIEW

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-unit-tests-zuul-docker-el7/3225/ : SUCCESS in 9m 13s
- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-opensource-repository-check/3985/ : SUCCESS in 11s

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-unit-tests-zuul-docker-el7/3277/ : SUCCESS in 10m 27s
- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-opensource-repository-check/4060/ : SUCCESS in 8s

**Comment by strouhaljan** [comment]
extended test - examples

**Comment by yenkins** [comment]
:arrow_right: Started Live Examples deploy to:
https://gdc-live-examples-914.herokuapp.com

It will be destroyed on merge/close. See [ci-infra](https://github.com/gooddata/ci-infra/blob/master/jenkins/jobs/templates/orchestrators/client-pr-dispatcher.groovy) and [gdc-ci.yaml](https://github.com/gooddata/gooddata-react-components/blob/master/gdc-ci.yaml).

**Comment by gdgate** [comment]
Build in pipeline check-extended aborted.

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-unit-tests-zuul-docker-el7/3279/ : SUCCESS in 13m 12s
- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-opensource-repository-check/4062/ : SUCCESS in 7s

**Comment by strouhaljan** [comment]
extended test - examples

**Comment by yenkins** [comment]
:arrow_right: Started Live Examples deploy to:
https://gdc-live-examples-914.herokuapp.com

It will be destroyed on merge/close. See [ci-infra](https://github.com/gooddata/ci-infra/blob/master/jenkins/jobs/templates/orchestrators/client-pr-dispatcher.groovy) and [gdc-ci.yaml](https://github.com/gooddata/gooddata-react-components/blob/master/gdc-ci.yaml).

File: src/components/visualizations/table/test/TableTransformation.spec.tsx [hunk 1]
```diff
@@ -61,7 +61,7 @@ describe("TableTransformation", () => {
                 drillableItems: [
                     { uri: "/uri" },
                     { uri: "identifier" },
-                    (_header: IMappingHeader, _afm: AFM.IAfm) => true,
+                    (_header: IMappingHeader, _afm: IHeaderPredicateContext) => true,
```
Comment (martinmosko) [inline src/components/visualizations/table/test/TableTransformation.spec.tsx]: _afm -> rename please

**Comment by gdgate** [comment]
Build succeeded (check-extended pipeline).

- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-pull-request-dispatcher/1103/ : SUCCESS in 29m 45s
- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-opensource-repository-check/4065/ : SUCCESS in 21s

File: src/helpers/CatalogHelper.ts [hunk 2]
```diff
@@ -67,27 +67,27 @@ export default class CatalogHelper {
     }
 
     public measure(name: string): string {
-        return get<CatalogHelper, string>(this, ["measures", name, "identifier"]);
+        return get(this, ["measures", name, "identifier"]);
```
Comment (martinmosko) [inline src/helpers/CatalogHelper.ts]: just wondering what is the motivation behind this changes?

File: src/components/visualizations/table/test/TableTransformation.spec.tsx (hunk 1 above)
Comment (strouhaljan) [reply src/components/visualizations/table/test/TableTransformation.spec.tsx]: Renamed to `_context`.

**Comment by strouhaljan** [comment]
extended test - storybook

**Comment by gdgate** [comment]
Build failed (check-extended pipeline).

- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-pull-request-dispatcher/1106/ : FAILURE in 6m 31s
- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-opensource-repository-check/4074/ : SUCCESS in 8s

**Comment by gdgate** [comment]
Build failed (check pipeline).

- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-unit-tests-zuul-docker-el7/3290/ : FAILURE in 7m 59s
- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-opensource-repository-check/4075/ : SUCCESS in 10s

File: src/helpers/CatalogHelper.ts (hunk 2 above)
Comment (strouhaljan) [reply src/helpers/CatalogHelper.ts]: We were unable to properly type these getters after upgrading TypeScript. The second parameter (key) should be one of the keys of `this`, but it is an array of strings in our case. After several discussions, we were unable to find solution.

File: src/helpers/sdkCache.ts [hunk 3]
```diff
@@ -34,7 +34,7 @@ export function clearSdkCache() {
  */
 export function getCachedOrLoad<T>(cacheKey: string, onCacheMiss: () => Promise<T>): Promise<T> {
     const sdkCache: ISdkCache = getSdkCache();
-    const cachedResponse = get<Promise<T>>(sdkCache, cacheKey);
+    const cachedResponse: Promise<T> = get<ISdkCache, keyof ISdkCache>(sdkCache, cacheKey);
```
[the remark this answers is not in the document]
Comment (strouhaljan) [reply src/helpers/sdkCache.ts]: Actually, this is correct according to the typing of `_.get`.

```
       /**
         * @param object The object to query.
         * @param path The path of the property to get.
         * @param defaultValue The value returned if the resolved value is undefined.
         * @return Returns the resolved value.
         */
        get<TObject extends object, TKey extends keyof TObject>(
            object: TObject,
            path: TKey | [TKey]
        ): TObject[TKey];
```

File: src/helpers/sorts.ts [hunk 4]
```diff
@@ -36,11 +36,11 @@ function getAttributeSortItems(
 
     const enrichedAggregation = aggregation
         ? {
-              aggregation: "sum",
+              aggregation: "sum" as "sum",
```
[the remark this answers is not in the document]
Comment (strouhaljan) [reply src/helpers/sorts.ts]: Fixed accordingly, there is no `base` and `enrichedAggregation`, just a single object.

File: src/index.ts [hunk 5]
```diff
@@ -59,12 +62,21 @@ import ColorUtils from "./components/visualizations/utils/color";
 import * as HeaderPredicateFactory from "./factory/HeaderPredicateFactory";
 import * as MappingHeader from "./interfaces/MappingHeader";
 
+interface ICoreComponents {
```
[the remark this answers is not in the document]
Comment (strouhaljan) [reply src/index.ts]: Move to `interfaces/CoreComponents.ts` file.

File: tsconfig.json [hunk 6]
```diff
@@ -14,12 +14,14 @@
     "noUnusedParameters": true,
     "outDir": "dist",
     "sourceMap": true,
-    "suppressImplicitAnyIndexErrors": true
+    "suppressImplicitAnyIndexErrors": true,
+    "allowSyntheticDefaultImports": true
```
[the remark this answers is not in the document]
Comment (strouhaljan) [reply tsconfig.json]: Once disabled, many of the used packages will throw an error while type checking because it has no default export:

- jest-mock
- jest-resolve
- jest-runtime
- events
- fs
- jest-haste-map

**Comment by gdgate** [comment]
Build in pipeline check aborted.

File: src/index.ts [hunk 7]
```diff
@@ -59,12 +62,21 @@ import ColorUtils from "./components/visualizations/utils/color";
...
+    BaseChart: React.ComponentClass<IBaseChartProps>;
+    Headline: React.ComponentClass<ICommonVisualizationProps & IDataSourceProviderInjectedProps>;
```
[the remark this answers is not in the document]
Comment (strouhaljan) [reply src/index.ts]: It should be ok because there is no need to explicitly expose our `visualizationLoadingHOC` wrapper component.

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-unit-tests-zuul-docker-el7/3310/ : SUCCESS in 15m 43s
- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-opensource-repository-check/4103/ : SUCCESS in 12s

File: src/helpers/tableCell.ts [hunk 8]
```diff
@@ -14,7 +14,8 @@ import {
 } from "../interfaces/Table";
 
 function getFormattedNumber(cellContent: MeasureCell, format: string, separators: ISeparators): string {
-    const parsedNumber: string | number = cellContent === null ? "" : parseFloat(cellContent);
+    const parsedNumber: string | number =
+        cellContent === null ? "" : typeof cellContent === "string" ? parseFloat(cellContent) : cellContent;
```
[the remark this answers is not in the document]
Comment (strouhaljan) [reply src/helpers/tableCell.ts]: There is no need for test coverage. `parseFloat` works with both `string` and `number` anyway.

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-unit-tests-zuul-docker-el7/3317/ : SUCCESS in 18m 40s
- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-opensource-repository-check/4112/ : SUCCESS in 9s

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-unit-tests-zuul-docker-el7/3329/ : SUCCESS in 9m 10s
- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-opensource-repository-check/4125/ : SUCCESS in 8s

**Comment by martinmosko** [comment]
LGTM

**Comment by gdgate** [comment]
Build failed (gate pipeline).

- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-storybook-zuul-docker/1782/ : ABORTED in 1h 00m 03s
- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-unit-tests-zuul-docker-el7/3342/ : SUCCESS in 17m 36s
- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-merge-compliance-check/914/ : SUCCESS in 27s
- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-examples-zuul-docker/963/ : SUCCESS in 34m 14s

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-unit-tests-zuul-docker-el7/3367/ : SUCCESS in 11m 55s
- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-opensource-repository-check/4166/ : SUCCESS in 8s

**Comment by martinmosko** [comment]
LGTM

**Comment by gdgate** [comment]
Build succeeded (gate pipeline).

- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-storybook-zuul-docker/1795/ : SUCCESS in 35m 46s
- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-unit-tests-zuul-docker-el7/3372/ : SUCCESS in 13m 24s
- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-merge-compliance-check/924/ : SUCCESS in 20s
- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-examples-zuul-docker/981/ : SUCCESS in 37m 37s
